torch/lib/THCS/generic/THCSTensorMath.cu

```diff
@@ -460,6 +462,115 @@ void THCSTensor_(cmul)(THCState *state, THCSTensor *r_, THCSTensor *t_, THCSTens
+              "Batch tensor Av expected, got %dD tensor", Av->nDimension);
+
+   long nBatch = THCTensor_(size)(state, Av, 0);
+   long nnz = THCTensor_(size)(state, Av, 1);
+   long m = Asz->data[0];
+   long n = Asz->data[1];
+
+   THCTensor_(resize2d)(state, rx_, nBatch, m);
+
+   THCIndexTensor *rowIndices = THCIndexTensor_(newSelect)(state, Ai, 0, 0);
+   THCIndexTensor *colIndices = THCIndexTensor_(newSelect)(state, Ai, 0, 1);
+   THCudaIntTensor *csr = THCSTensor_(toCSR)(state, rowIndices, m, nnz);
+   THCudaIntTensor *colIndicesInt = THCudaIntTensor_newWithSize1d(state, colIndices->size[0]);
+   THCudaIntTensor_copyCudaLong(state, colIndicesInt, colIndices);
+
+   cusparseStatus_t cusparse_status;
+   cusolverStatus_t cusolver_status;
+   cudaError_t cuda_status;
+   csrqrInfo_t info;
+
+   cusolverSpHandle_t cusolverH;
+   cusolver_status = cusolverSpCreate(&cusolverH);
+   THAssert(CUSOLVER_STATUS_SUCCESS == cusolver_status);
+
+   cusparseMatDescr_t descrA;
+   size_t size_qr = 0;
+   size_t size_internal = 0;
+   void *buffer_qr = NULL;
+
+   cusparse_status = cusparseCreateMatDescr(&descrA);
+   THAssert(cusparse_status == CUSPARSE_STATUS_SUCCESS);
+
+   cusparseSetMatType(descrA, CUSPARSE_MATRIX_TYPE_GENERAL);
+#if TH_INDEX_BASE == 0
+   cusparseSetMatIndexBase(descrA, CUSPARSE_INDEX_BASE_ZERO);
+#elif TH_INDEX_BASE ==1
+   cusparseSetMatIndexBase(descrA, CUSPARSE_INDEX_BASE_ONE);
+#else
+     THError("Unknown TH_INDEX_BASE.");
+#endif
+
+   cusolver_status = cusolverSpCreateCsrqrInfo(&info);
+   THAssert(cusolver_status == CUSOLVER_STATUS_SUCCESS);
+
+   cusolver_status = cusolverSpXcsrqrAnalysisBatched(
+       cusolverH, m, n, nnz,
+       descrA,
+       THCudaIntTensor_data(state, csr),
+       THCudaIntTensor_data(state, colIndicesInt),
+       info);
+   THAssert(cusolver_status == CUSOLVER_STATUS_SUCCESS);
+
+   cusolver_status =
+#if defined(THCS_REAL_IS_FLOAT)
+     cusolverSpScsrqrBufferInfoBatched(
+#elif defined(THCS_REAL_IS_DOUBLE)
+     cusolverSpDcsrqrBufferInfoBatched(
+#endif
+       cusolverH, m, n, nnz,
+       descrA,
+       THCTensor_(data)(state, Av),
+       THCudaIntTensor_data(state, csr),
+       THCudaIntTensor_data(state, colIndicesInt),
+       nBatch,
+       info,
+       &size_internal,
+       &size_qr);
+   THAssert(cusolver_status == CUSOLVER_STATUS_SUCCESS);
+
+   cuda_status = cudaMalloc((void**)&buffer_qr, size_qr);
+   THAssert(cuda_status == cudaSuccess);
+
+   cusolver_status =
+#if defined(THCS_REAL_IS_FLOAT)
+     cusolverSpScsrqrsvBatched(
+#elif defined(THCS_REAL_IS_DOUBLE)
+     cusolverSpDcsrqrsvBatched(
+#endif
+       cusolverH, m, n, nnz,
+       descrA,
+       THCTensor_(data)(state, Av),
+       THCudaIntTensor_data(state, csr),
+       THCudaIntTensor_data(state, colIndicesInt),
+       THCTensor_(data)(state, b),
+       THCTensor_(data)(state, rx_),
+       nBatch, info,
+       buffer_qr);
+   THAssert(cusolver_status == CUSOLVER_STATUS_SUCCESS);
```

Comment:
It'd be nice to add macros for checking these error codes that would also print readable messages